src/Microsoft.OData.Core/RawValueWriter.cs

```diff
@@ -154,6 +153,86 @@ internal void Flush()
             }
         }
 
+        /// <summary>
+        /// Asynchronously start writing a raw output. This should only be called once.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal async Task StartAsync()
+        {
+            if (this.settings.HasJsonPaddingFunction())
+            {
+                await this.textWriter.WriteAsync(this.settings.JsonPCallback)
+                    .ConfigureAwait(false);
+                await this.textWriter.WriteAsync(JsonConstants.StartPaddingFunctionScope)
+                    .ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously end the writing of a raw output. This should be the last thing called.
+        /// </summary>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal Task EndAsync()
+        {
+            if (this.settings.HasJsonPaddingFunction())
+            {
+                return this.textWriter.WriteAsync(JsonConstants.EndPaddingFunctionScope);
+            }
+
+            return TaskUtils.CompletedTask;
+        }
+
+        /// <summary>
+        /// Asynchronously converts the specified <paramref name="value"/> into its raw format and writes it to the output.
+        /// The value has to be of enumeration or primitive type. Only one WriteRawValue call should be made before this object gets disposed.
+        /// </summary>
+        /// <param name="value">The (non-binary) value to write.</param>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        /// <remarks>We do not accept binary values here; WriteBinaryValue should be used for binary data.</remarks>
+        /// <returns>A task that represents the asynchronous write operation.</returns>
+        internal Task WriteRawValueAsync(object value)
+        {
+            Debug.Assert(!(value is byte[]), "!(value is byte[])");
+
+            string valueAsString;
+            if (value is ODataEnumValue enumValue)
+            {
+                return this.textWriter.WriteAsync(enumValue.Value);
+            }
+            else if (value is Geometry || value is Geography)
+            {
+                return TaskUtils.GetTaskForSynchronousOperation(
+                    () => PrimitiveConverter.Instance.WriteJsonLight(value, jsonWriter));

```

Comment:
Out of curiosity, does this `WriteJsonLight` method write to a stream (the `jsonWriter`)? If so, why doesn't it have an async equivalent?

Reply:
It was a deliberate decision to retain asynchronous wrappers over synchronous methods every place where `PrimitiveConverter.Instance.WriteJsonLight(object, IJsonWriter)` method has been used. If you explore deeper, you'll realize that this method is used for writing spatial data. Since extending asynchronous support to Microsoft.Spatial library at this time would have resulted into an explosion of scope, wrapping the synchronous call was the alternative option. You'll find that there's a test to verify that it works as expected.